src/extensions/metadata.rs

```diff
@@ -0,0 +1,166 @@
+//! Adds support for the IMAP ENABLE command specificed in [RFC
+//! 5464](https://tools.ietf.org/html/rfc5464).
+
+use crate::client::*;
+use crate::error::{Error, ParseError, Result};
+use crate::types::*;
+use imap_proto::types::{MailboxDatum, Metadata, Response};
+use std::io::{Read, Write};
+use std::sync::mpsc;
+
+trait CmdListItemFormat {
+    fn format_as_cmd_list_item(&self) -> String;
+}
+
+impl CmdListItemFormat for Metadata {
+    fn format_as_cmd_list_item(&self) -> String {
+        format!(
+            "{} {}",
+            validate_str(self.entry.as_str()).unwrap(),
+            self.value
+                .as_ref()
+                .map(|v| validate_str(v.as_str()).unwrap())
+                .unwrap_or("NIL".to_string())
+        )
+    }
+}
+
+/// Represents variants of DEPTH parameters for GETMETADATA command
+#[derive(Debug, Copy, Clone)]
+pub enum MetadataDepth {
+    /// Depth 0 for get metadata
+    Zero,
+    /// Depth 1 for get metadata
+    One,
+    /// Depth infinity for get metadata
+    Inf,
+}
+
+impl MetadataDepth {
+    fn depth_str<'a>(self) -> &'a str {
+        match self {
+            MetadataDepth::Zero => return "0",
+            MetadataDepth::One => return "1",
+            MetadataDepth::Inf => return "infinity",
+        }
+    }
+}
+
+fn parse_metadata<'a>(
+    mut lines: &'a [u8],
+    unsolicited: &'a mut mpsc::Sender<UnsolicitedResponse>,
+) -> Result<Vec<Metadata>> {
+    let mut res: Vec<Metadata> = Vec::new();
+    loop {
+        if lines.is_empty() {
+            break Ok(res);
+        }
+
+        match imap_proto::parse_response(lines) {
+            Ok((rest, resp)) => {
+                lines = rest;
+                match resp {
+                    Response::MailboxData(MailboxDatum::MetadataSolicited {
+                        mailbox: _,
+                        mut values,
+                    }) => {
+                        res.append(&mut values);
+                    }
+                    Response::MailboxData(MailboxDatum::MetadataUnsolicited {
```

Comment:
This doesn't quite seem right. I think this method should look similar to the other parsers, where it _just_ handles the solicited responses (here `MailboxDatum::MetadataSolicited`), and then has a fallback clause to catch _any_ unilateral server response:
https://github.com/jonhoo/rust-imap/blob/fbdb5d3be4465d3b010369a9b98ad8b0631b16fa/src/parse.rs#L183-L188

The reason for this is that unilateral responses may come _at any time_, including in the middle of other commands. So even while you are running the `GETMETADATA` command, a unilateral `EXPUNGE` response may arrive.

Reply:
The code for handling `MetadataUnsolicited` should be moved into `handle_unilateral`.